osm2streets/src/paint.rs

```diff
@@ -0,0 +1,235 @@
+use crate::{marking, marking::Marking};
+
+// We use geom and stay in map space. Output is done in latlon.
+use geom::{Angle, Distance, Line, PolyLine, Polygon, Pt2D, Ring};
+
+use crate::lanes::TrafficMode;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct PaintArea {
+    // Because I'm lazy and don't want to make different "map space" and "lonlat space" PaintArea
+    // types, I'm using geo::Polygon, so we can just swap out the coords in place. Not ideal.
+    /// A simple ring.
+    pub area: geo::Polygon,
+    pub color: PaintColor,
+}
+impl PaintArea {
+    pub fn new(area: Ring, color: PaintColor) -> Self {
+        Self {
+            area: area.into_polygon().into(),
+            color,
+        }
+    }
+
+    pub fn from(area: Ring) -> Self {
+        Self {
+            area: area.into_polygon().into(),
+            color: PaintColor::White,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PaintColor {
+    White,
+    Yellow,
+}
+
+impl PaintColor {
+    pub fn to_str(&self) -> &str {
+        match self {
+            Self::White => "white",
+            Self::Yellow => "yellow",
+        }
+    }
+}
+
+trait Paint<T> {
+    fn paint(&self, geometry: &T) -> Vec<PaintArea>;
+}
+
+impl Marking {
+    pub fn paint(&self) -> Vec<PaintArea> {
+        match self {
+            Marking::Longitudinal(g, m) => m.paint(g),
+            Marking::Transverse(g, m) => m.paint(g),
+            Marking::Symbol(g0, g1, m) => m.paint(&(g0, g1)),
+            Marking::Area(g, m) => m.paint(g),
+        }
+    }
+}
+
+// Measurements for the default non-locale rendering of markings.
+const LINE_WIDTH: Distance = Distance::const_meters(0.10);
+const LINE_WIDTH_THIN: Distance = Distance::const_meters(0.08);
+const LINE_WIDTH_THICK: Distance = Distance::const_meters(0.30);
+const DASH_LENGTH_SHORT: Distance = Distance::const_meters(1.0);
+const DASH_GAP_SHORT: Distance = Distance::const_meters(1.0);
+const DASH_LENGTH_LONG: Distance = Distance::const_meters(2.0);
+const DASH_GAP_LONG: Distance = Distance::const_meters(4.5);
+
+impl Paint<PolyLine> for marking::Longitudinal {
+    fn paint(&self, separator: &PolyLine) -> Vec<PaintArea> {
+        // TODO incorporate colors throughout instead of only collecting rings:
```

Comment:
Maybe something like the [GeomBatch API](https://github.com/a-b-street/abstreet/blob/f42aaafb24b3533b63bf1cedb747dfaa95bbf4b9/widgetry/src/geom/mod.rs#L38) would be helpful eventually